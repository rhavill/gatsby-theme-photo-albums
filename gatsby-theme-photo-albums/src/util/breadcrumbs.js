import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs