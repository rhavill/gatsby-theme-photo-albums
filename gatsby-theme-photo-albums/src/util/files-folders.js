const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}