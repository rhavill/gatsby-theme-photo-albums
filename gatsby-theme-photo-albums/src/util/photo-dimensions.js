export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}