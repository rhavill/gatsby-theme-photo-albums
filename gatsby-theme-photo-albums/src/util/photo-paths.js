const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => indexedForEach(
    (file, i) => {
      const pageNumber = Math.ceil((i + 1) / photosPerPage)
      photoPaths[file.url] = getPhotoPathWithPage(
        pageNumber, file.url
      )
    }, 
    groupedFiles[relativeDirectory]
  ), keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

module.exports = {getPhotoPathsWithPages, getPhotoPathWithPage}