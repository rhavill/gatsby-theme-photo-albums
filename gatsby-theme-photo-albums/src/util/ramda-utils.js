const addIndex = require('ramda/src/addIndex')
const complement = require('ramda/src/complement')
const compose = require('ramda/src/compose')
const forEach = require('ramda/src/forEach')
const groupBy = require('ramda/src/groupBy')
const isNil = require('ramda/src/isNil')
const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
const objectArrayToPropArray = (property, objects) => map(prop(property), objects)

const mapIndexed = addIndex(map)

const indexedForEach = addIndex(forEach)

const groupByProp = property => compose(groupBy, prop)(property)

const isNotNil = complement(isNil)

module.exports = {
  objectArrayToPropArray,
  mapIndexed,
  indexedForEach,
  groupByProp,
  isNotNil,
}
