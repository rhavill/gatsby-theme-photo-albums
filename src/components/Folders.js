import React from 'react'
import PropTypes from 'prop-types'
import addIndex from 'ramda/src/addIndex'
import compose from 'ramda/src/compose'
import map from 'ramda/src/map'
import Folder from './Folder'
import getChildPaths from '../util/files-folders'

const mapIndexed = addIndex(map)

const Folders = ({path, folders, icon}) => {
  return compose(
    mapIndexed((folder, i) => 
      <Folder key={i} path={folder} icon={icon} />
    ),
    getChildPaths(path)
  )(folders)
}

Folders.propTypes = {
  path: PropTypes.string.isRequired,
  folders: PropTypes.arrayOf(
    PropTypes.string.isRequired, 
  ).isRequired,
  icon: PropTypes.shape({
    src: PropTypes.string.isRequired,
    srcSet: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired
}

export default Folders