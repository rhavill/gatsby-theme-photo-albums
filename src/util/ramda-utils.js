const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
exports.objectArrayToPropArray =  (property, objects) => map(prop(property), objects)
