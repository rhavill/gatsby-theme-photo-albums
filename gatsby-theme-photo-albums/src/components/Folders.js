import React from 'react'
import PropTypes from 'prop-types'
import Folder from './Folder'
import {mapIndexed} from '../util/ramda-utils'

const Folders = ({folders, icon}) => {
  return mapIndexed((folder, i) => 
    <Folder key={i} path={folder} icon={icon} />
  )(folders)
}

Folders.propTypes = {
  folders: PropTypes.arrayOf(
    PropTypes.string.isRequired, 
  ).isRequired,
  icon: PropTypes.shape({
    src: PropTypes.string.isRequired,
    srcSet: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired
}

export default Folders