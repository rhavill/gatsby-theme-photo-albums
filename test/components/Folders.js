import React from 'react'
import renderer from 'react-test-renderer'
import {cleanup, render} from '@testing-library/react'
import Folders from '../../src/components/Folders'
import {objectArrayToPropArray} from '../../src/util/ramda-utils'
import folderData from '../../test-data/source-filesystem-folder-data'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

const folders = objectArrayToPropArray('url', folderData.folders.nodes)
const icon =folderData.folderIcon.childImageSharp.fixed

describe('Folders', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Folders path='/base/level-one' folders={folders} icon={icon} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('displays the correct number of folder icons', () => {
    const folders = [
      '/base/level-one/level-two',
      '/base/level-one/another-level-two',
    ]
    const {getAllByTestId} = render(
      <Folders path='/base/level-one' folders={folders} icon={icon} />
    )
    expect(getAllByTestId('folder').length).toBe(2)
  })
})