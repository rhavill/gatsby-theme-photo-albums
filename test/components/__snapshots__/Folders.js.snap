// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Folders renders correctly 1`] = `
Array [
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="/base/"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="/base/"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"/base/\\" title=\\"/base/\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        /base/
      </div>
    </a>
  </article>,
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/level-one"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="Level One"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="Level One"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"Level One\\" title=\\"Level One\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        Level One
      </div>
    </a>
  </article>,
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/another-level-one"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="Another Level One"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="Another Level One"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"Another Level One\\" title=\\"Another Level One\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        Another Level One
      </div>
    </a>
  </article>,
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/yet-another-level-one"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="Yet Another Level One"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="Yet Another Level One"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"Yet Another Level One\\" title=\\"Yet Another Level One\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        Yet Another Level One
      </div>
    </a>
  </article>,
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/level-one/level-two"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="Level Two"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="Level Two"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"Level Two\\" title=\\"Level Two\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        Level Two
      </div>
    </a>
  </article>,
  <article
    className="folder"
    data-testid="folder"
  >
    <a
      href="/base/level-one/another-level-two"
    >
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 250,
            "overflow": "hidden",
            "position": "relative",
            "width": 250,
          }
        }
      >
        <picture>
          <source
            srcSet="/static/hash/f17d3/folder.png 1x"
          />
          <img
            alt="Another Level Two"
            height={250}
            loading="lazy"
            onLoad={[Function]}
            src="/static/hash/f17d3/folder.png"
            srcSet="/static/hash/f17d3/folder.png 1x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            title="Another Level Two"
            width={250}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/hash/f17d3/folder.png 1x\\" /><img loading=\\"lazy\\" width=\\"250\\" height=\\"250\\" srcset=\\"/static/hash/f17d3/folder.png 1x\\" src=\\"/static/hash/f17d3/folder.png\\" alt=\\"Another Level Two\\" title=\\"Another Level Two\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="folder-title"
      >
        Another Level Two
      </div>
    </a>
  </article>,
]
`;
